File: packages/vault/src/test/plaintext-vault.test.ts
import * as os from "os";
import * as path from "path";
import * as fs from "fs/promises";
import { PlaintextVault } from "../lib/plaintext-vault";
import { ISecret } from "../lib/vault";

//
// Creates a unique temporary directory for each test so that tests are
// fully isolated from one another and from the real ~/.config/vault.
//
async function makeTempDir(): Promise<string> {
    return fs.mkdtemp(path.join(os.tmpdir(), "vault-test-"));
}

//
// Removes the temporary directory and all its contents after a test.
//
async function removeTempDir(dirPath: string): Promise<void> {
    await fs.rm(dirPath, { recursive: true, force: true });
}

describe("PlaintextVault", () => {
    let tempDir: string;
    let vault: PlaintextVault;

    beforeEach(async () => {
        tempDir = await makeTempDir();
        vault = new PlaintextVault(tempDir);
    });

    afterEach(async () => {
        await removeTempDir(tempDir);
    });

    describe("get", () => {
        test("returns undefined for a secret that does not exist", async () => {
            const result = await vault.get("nonexistent");
            expect(result).toBeUndefined();
        });

        test("returns the secret after it has been set", async () => {
            const secret: ISecret = { name: "my-key", type: "api-key", value: "abc123" };
            await vault.set(secret);
            const result = await vault.get("my-key");
            expect(result).toEqual(secret);
        });

        test("returns the latest value when a secret is overwritten", async () => {
            await vault.set({ name: "token", type: "api-key", value: "old-value" });
            await vault.set({ name: "token", type: "api-key", value: "new-value" });
            const result = await vault.get("token");
            expect(result?.value).toBe("new-value");
        });

        test("returns undefined when the vault file exists but holds no such name", async () => {
            await vault.set({ name: "present", type: "password", value: "here" });
            const result = await vault.get("absent");
            expect(result).toBeUndefined();
        });
    });

    describe("set", () => {
        test("creates the vault directory if it does not exist", async () => {
            const nestedDir = path.join(tempDir, "nested", "vault");
            const nestedVault = new PlaintextVault(nestedDir);
            await nestedVault.set({ name: "key", type: "password", value: "secret" });
            const result = await nestedVault.get("key");
            expect(result?.value).toBe("secret");
        });

        test("stores the name, type, and value of the secret", async () => {
            const secret: ISecret = { name: "db-pass", type: "password", value: "hunter2" };
            await vault.set(secret);
            const result = await vault.get("db-pass");
            expect(result?.name).toBe("db-pass");
            expect(result?.type).toBe("password");
            expect(result?.value).toBe("hunter2");
        });

        test("supports arbitrary type strings", async () => {
            const secret: ISecret = { name: "cred", type: "s3-credentials", value: '{"accessKeyId":"AKIA...","secretAccessKey":"xyz"}' };
            await vault.set(secret);
            const result = await vault.get("cred");
            expect(result?.type).toBe("s3-credentials");
        });

        test("creates the vault file when it does not exist yet", async () => {
            await vault.set({ name: "first", type: "password", value: "one" });
            const stats = await fs.stat(path.join(tempDir, "vault.json"));
            expect(stats.isFile()).toBe(true);
        });

        test("preserves the secrets already held in the vault file", async () => {
            await vault.set({ name: "first", type: "password", value: "one" });
            await vault.set({ name: "second", type: "api-key", value: "two" });
            await vault.set({ name: "third", type: "password", value: "three" });
            const first = await vault.get("first");
            const second = await vault.get("second");
            const third = await vault.get("third");
            expect(first?.value).toBe("one");
            expect(second?.value).toBe("two");
            expect(third?.value).toBe("three");
        });

        test("supports a key pair secret with both private and public key in the value", async () => {
            const keyPairValue = JSON.stringify({ privateKey: "-----BEGIN PRIVATE KEY-----", publicKey: "-----BEGIN PUBLIC KEY-----" });
            await vault.set({ name: "my-keypair", type: "key-pair", value: keyPairValue });
            const result = await vault.get("my-keypair");
            expect(result?.type).toBe("key-pair");
            const parsed = JSON.parse(result!.value);
            expect(parsed.privateKey).toBe("-----BEGIN PRIVATE KEY-----");
            expect(parsed.publicKey).toBe("-----BEGIN PUBLIC KEY-----");
        });
    });

    describe("list", () => {
        test("returns an empty array when the vault is empty", async () => {
            const secrets = await vault.list();
            expect(secrets).toEqual([]);
        });

        test("returns an empty array when the vault file does not exist", async () => {
            const emptyVault = new PlaintextVault(path.join(tempDir, "does-not-exist"));
            const secrets = await emptyVault.list();
            expect(secrets).toEqual([]);
        });

        test("returns all stored secrets", async () => {
            await vault.set({ name: "alpha", type: "password", value: "aaa" });
            await vault.set({ name: "beta", type: "api-key", value: "bbb" });
            const secrets = await vault.list();
            const names = secrets.map(secret => secret.name).sort();
            expect(names).toEqual(["alpha", "beta"]);
        });

        test("does not include deleted secrets", async () => {
            await vault.set({ name: "keep", type: "password", value: "keep-value" });
            await vault.set({ name: "remove", type: "password", value: "remove-value" });
            await vault.delete("remove");
            const secrets = await vault.list();
            expect(secrets).toHaveLength(1);
            expect(secrets[0].name).toBe("keep");
        });

        test("ignores a stray file sitting alongside the vault file", async () => {
            await vault.set({ name: "real", type: "api-key", value: "val" });
            // Write a stray file that should be ignored.
            await fs.writeFile(path.join(tempDir, "noise.txt"), "ignore me", "utf8");
            const secrets = await vault.list();
            expect(secrets).toHaveLength(1);
        });
    });

    describe("a malformed vault file", () => {
        //
        // Writes a vault file that is not valid JSON, so the vault has to
        // fail loudly rather than reporting an empty vault and then
        // overwriting whatever was really in there.
        //
        async function writeMalformedVaultFile(): Promise<void> {
            await fs.writeFile(path.join(tempDir, "vault.json"), "{ this is not json", "utf8");
        }

        test("makes get throw", async () => {
            await writeMalformedVaultFile();
            await expect(vault.get("anything")).rejects.toThrow();
        });

        test("makes list throw", async () => {
            await writeMalformedVaultFile();
            await expect(vault.list()).rejects.toThrow();
        });
    });

    describe("delete", () => {
        test("removes a secret so that get returns undefined afterwards", async () => {
            await vault.set({ name: "gone", type: "password", value: "byebye" });
            await vault.delete("gone");
            const result = await vault.get("gone");
            expect(result).toBeUndefined();
        });

        test("does nothing when the secret does not exist", async () => {
            await expect(vault.delete("no-such-secret")).resolves.toBeUndefined();
        });
    });

    describe("secret names with special characters", () => {
        test("handles names containing spaces", async () => {
            await vault.set({ name: "my secret", type: "password", value: "spaced" });
            const result = await vault.get("my secret");
            expect(result?.value).toBe("spaced");
        });

        test("handles names containing slashes", async () => {
            await vault.set({ name: "org/repo/token", type: "api-key", value: "tok" });
            const result = await vault.get("org/repo/token");
            expect(result?.value).toBe("tok");
        });

        test("handles names containing unicode characters", async () => {
            await vault.set({ name: "clé-secrète", type: "password", value: "motdepasse" });
            const result = await vault.get("clé-secrète");
            expect(result?.value).toBe("motdepasse");
        });

        test("handles multiple specially named secrets without collision", async () => {
            await vault.set({ name: "a/b", type: "password", value: "slash" });
            await vault.set({ name: "a%2Fb", type: "password", value: "percent" });
            const slash = await vault.get("a/b");
            const percent = await vault.get("a%2Fb");
            expect(slash?.value).toBe("slash");
            expect(percent?.value).toBe("percent");
        });
    });

    describe("file permissions", () => {
        // Permission checks are only meaningful on POSIX platforms.
        const isWindows = process.platform === "win32";

        test("vault file is created with owner-only permissions (0o600)", async () => {
            if (isWindows) {
                return;
            }
            await vault.set({ name: "perm-test", type: "password", value: "s3cr3t" });
            const filePath = path.join(tempDir, "vault.json");
            const stats = await fs.stat(filePath);
            const mode = stats.mode & 0o777;
            expect(mode).toBe(0o600);
        });

        test("vault directory is created with owner-only permissions (0o700)", async () => {
            if (isWindows) {
                return;
            }
            const newVaultDir = path.join(tempDir, "new-vault");
            const newVault = new PlaintextVault(newVaultDir);
            await newVault.set({ name: "key", type: "password", value: "val" });
            const stats = await fs.stat(newVaultDir);
            const mode = stats.mode & 0o777;
            expect(mode).toBe(0o700);
        });
    });

    describe("exists", () => {
        test("returns false before any secrets are stored", () => {
            const freshVault = new PlaintextVault(path.join(tempDir, "fresh"));
            expect(freshVault.exists()).toBe(false);
        });

        test("returns false when the vault directory exists but holds no vault file", () => {
            expect(vault.exists()).toBe(false);
        });

        test("returns true after a secret has been stored", async () => {
            await vault.set({ name: "anything", type: "password", value: "val" });
            expect(vault.exists()).toBe(true);
        });
    });
});

File: packages/vault-zig/src/test/plaintext-vault.test.zig
const std = @import("std");
const vault_zig = @import("vault-zig");
const PlaintextVault = vault_zig.plaintext_vault.PlaintextVault;
const ISecret = vault_zig.vault.ISecret;

//
// A unique temporary directory for one test, so that tests are fully isolated from one another
// and from the real ~/.config/photosphere/vault.
//
const TempDir = struct {
    // The directory created by std.testing.tmpDir.
    tmp_dir: std.testing.TmpDir,

    // The path of the directory relative to the working directory.
    path: []const u8,

    //
    // Removes the temporary directory and all its contents after a test.
    //
    fn remove(self: *TempDir) void {
        self.tmp_dir.cleanup();
    }
};

//
// Creates a unique temporary directory for a test.
//
fn makeTempDir(allocator: std.mem.Allocator) !TempDir {
    const tmp_dir = std.testing.tmpDir(.{ .iterate = true });
    const path = try std.fmt.allocPrint(allocator, ".zig-cache/tmp/{s}", .{tmp_dir.sub_path});
    return .{ .tmp_dir = tmp_dir, .path = path };
}

//
// Returns the Unix permission bits of a file or directory.
//
fn permissionBits(io: std.Io, path: []const u8) !u32 {
    const stat = try std.Io.Dir.cwd().statFile(io, path, .{});
    return @intCast(@intFromEnum(stat.permissions) & 0o777);
}

//
// Orders secrets by name.
//
fn secretNameLessThan(context: void, left: ISecret, right: ISecret) bool {
    _ = context;
    return std.mem.order(u8, left.name, right.name) == .lt;
}

test "get: returns undefined for a secret that does not exist" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    const result = try vault.get(allocator, std.testing.io, "nonexistent");
    try std.testing.expect(result == null);
}

test "get: returns the secret after it has been set" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    const secret: ISecret = .{ .name = "my-key", .type = "api-key", .value = "abc123" };
    try vault.set(allocator, io, secret);
    const result = (try vault.get(allocator, io, "my-key")).?;
    try std.testing.expectEqualStrings(secret.name, result.name);
    try std.testing.expectEqualStrings(secret.type, result.type);
    try std.testing.expectEqualStrings(secret.value, result.value);
}

test "get: returns the latest value when a secret is overwritten" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "token", .type = "api-key", .value = "old-value" });
    try vault.set(allocator, io, .{ .name = "token", .type = "api-key", .value = "new-value" });
    const result = (try vault.get(allocator, io, "token")).?;
    try std.testing.expectEqualStrings("new-value", result.value);
}

test "set: creates the vault directory if it does not exist" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();

    const nestedDir = try std.fs.path.join(allocator, &.{ temp_dir.path, "nested", "vault" });
    var nestedVault = PlaintextVault.init(nestedDir);
    try nestedVault.set(allocator, io, .{ .name = "key", .type = "password", .value = "secret" });
    const result = (try nestedVault.get(allocator, io, "key")).?;
    try std.testing.expectEqualStrings("secret", result.value);
}

test "set: stores the name, type, and value of the secret" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "db-pass", .type = "password", .value = "hunter2" });
    const result = (try vault.get(allocator, io, "db-pass")).?;
    try std.testing.expectEqualStrings("db-pass", result.name);
    try std.testing.expectEqualStrings("password", result.type);
    try std.testing.expectEqualStrings("hunter2", result.value);
}

test "set: supports arbitrary type strings" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "cred", .type = "s3-credentials", .value = "{\"accessKeyId\":\"AKIA...\",\"secretAccessKey\":\"xyz\"}" });
    const result = (try vault.get(allocator, io, "cred")).?;
    try std.testing.expectEqualStrings("s3-credentials", result.type);
}

test "set: supports a key pair secret with both private and public key in the value" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    const keyPairValue = "{\"privateKey\":\"-----BEGIN PRIVATE KEY-----\",\"publicKey\":\"-----BEGIN PUBLIC KEY-----\"}";
    try vault.set(allocator, io, .{ .name = "my-keypair", .type = "key-pair", .value = keyPairValue });
    const result = (try vault.get(allocator, io, "my-keypair")).?;
    try std.testing.expectEqualStrings("key-pair", result.type);
    const parsed = try std.json.parseFromSliceLeaky(std.json.Value, allocator, result.value, .{});
    try std.testing.expectEqualStrings("-----BEGIN PRIVATE KEY-----", parsed.object.get("privateKey").?.string);
    try std.testing.expectEqualStrings("-----BEGIN PUBLIC KEY-----", parsed.object.get("publicKey").?.string);
}

test "list: returns an empty array when the vault is empty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    const secrets = try vault.list(allocator, std.testing.io);
    try std.testing.expectEqual(@as(usize, 0), secrets.len);
}

test "list: returns an empty array when the vault directory does not exist" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();

    var emptyVault = PlaintextVault.init(try std.fs.path.join(allocator, &.{ temp_dir.path, "does-not-exist" }));
    const secrets = try emptyVault.list(allocator, std.testing.io);
    try std.testing.expectEqual(@as(usize, 0), secrets.len);
}

test "list: returns all stored secrets" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "beta", .type = "api-key", .value = "bbb" });
    try vault.set(allocator, io, .{ .name = "alpha", .type = "password", .value = "aaa" });
    const secrets = try vault.list(allocator, io);
    try std.testing.expectEqual(@as(usize, 2), secrets.len);

    // Entries are returned in file name order (like Node's sorted fs.readdir).
    try std.testing.expectEqualStrings("alpha", secrets[0].name);
    try std.testing.expectEqualStrings("beta", secrets[1].name);
}

// Not ported: "does not include deleted secrets" (PlaintextVault.delete is not ported: not used by psi replicate or psi verify).

test "list: ignores files without the .json extension" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "real", .type = "api-key", .value = "val" });

    // Write a stray file that should be ignored.
    try temp_dir.tmp_dir.dir.writeFile(io, .{ .sub_path = "noise.txt", .data = "ignore me" });
    const secrets = try vault.list(allocator, io);
    try std.testing.expectEqual(@as(usize, 1), secrets.len);
}

// Not ported: "delete" tests (PlaintextVault.delete is not ported: not used by psi replicate or psi verify).

test "secret names with special characters: handles names containing spaces" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "my secret", .type = "password", .value = "spaced" });
    const result = (try vault.get(allocator, io, "my secret")).?;
    try std.testing.expectEqualStrings("spaced", result.value);
}

test "secret names with special characters: handles names containing slashes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "org/repo/token", .type = "api-key", .value = "tok" });
    const result = (try vault.get(allocator, io, "org/repo/token")).?;
    try std.testing.expectEqualStrings("tok", result.value);
}

test "secret names with special characters: handles names containing unicode characters" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "clé-secrète", .type = "password", .value = "motdepasse" });
    const result = (try vault.get(allocator, io, "clé-secrète")).?;
    try std.testing.expectEqualStrings("motdepasse", result.value);
}

test "secret names with special characters: handles multiple specially named secrets without collision" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "a/b", .type = "password", .value = "slash" });
    try vault.set(allocator, io, .{ .name = "a%2Fb", .type = "password", .value = "percent" });
    const slash = (try vault.get(allocator, io, "a/b")).?;
    const percent = (try vault.get(allocator, io, "a%2Fb")).?;
    try std.testing.expectEqualStrings("slash", slash.value);
    try std.testing.expectEqualStrings("percent", percent.value);
}

test "file permissions: secret file is created with owner-only permissions (0o600)" {
    if (@import("builtin").os.tag == .windows) {
        return;
    }
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    try vault.set(allocator, io, .{ .name = "perm-test", .type = "password", .value = "s3cr3t" });
    const filePath = try std.fs.path.join(allocator, &.{ temp_dir.path, "perm-test.json" });
    try std.testing.expectEqual(@as(u32, 0o600), try permissionBits(io, filePath));
}

test "file permissions: vault directory is created with owner-only permissions (0o700)" {
    if (@import("builtin").os.tag == .windows) {
        return;
    }
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();

    const newVaultDir = try std.fs.path.join(allocator, &.{ temp_dir.path, "new-vault" });
    var newVault = PlaintextVault.init(newVaultDir);
    try newVault.set(allocator, io, .{ .name = "key", .type = "password", .value = "val" });
    try std.testing.expectEqual(@as(u32, 0o700), try permissionBits(io, newVaultDir));
}

// Not ported: "exists" tests (PlaintextVault.exists is not ported: not used by psi replicate or psi verify).

test "encodeSecretName percent-encodes like encodeURIComponent" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const encode = vault_zig.plaintext_vault.encodeSecretName;

    try std.testing.expectEqualStrings("default%3As3", try encode(allocator, "default:s3"));
    try std.testing.expectEqualStrings("a%252Fb", try encode(allocator, "a%2Fb"));
    try std.testing.expectEqualStrings("cl%C3%A9", try encode(allocator, "clé"));
    try std.testing.expectEqualStrings("AZaz09-_.!~*'()", try encode(allocator, "AZaz09-_.!~*'()"));
    try std.testing.expectEqualStrings("%20%2F%3F%23%26%3D%2B%24%2C%40%5B%5D", try encode(allocator, " /?#&=+$,@[]"));
}

test "decodeSecretName reverses encodeSecretName and rejects malformed input" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const decode = vault_zig.plaintext_vault.decodeSecretName;

    try std.testing.expectEqualStrings("default:s3", try decode(allocator, "default%3As3"));
    try std.testing.expectEqualStrings("clé", try decode(allocator, "cl%C3%A9"));
    try std.testing.expectEqualStrings("a%2Fb", try decode(allocator, "a%252Fb"));
    try std.testing.expectError(error.Thrown, decode(allocator, "bad%2"));
    try std.testing.expectEqualStrings("URI malformed", @import("utils-zig").errors.lastErrorMessage());
    try std.testing.expectError(error.Thrown, decode(allocator, "bad%zz"));
    try std.testing.expectError(error.Thrown, decode(allocator, "bad%C3"));
}

test "DEFAULT_VAULT_DIR is ~/.config/photosphere/vault" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const process_env = @import("node-utils-zig").process_env;

    var environ_map = std.process.Environ.Map.init(allocator);
    try environ_map.put("HOME", "/home/tester");
    try environ_map.put("USERPROFILE", "/home/tester");
    process_env.setEnvironMap(&environ_map);
    defer process_env.setEnvironMap(null);

    const expected = try std.fs.path.join(allocator, &.{ "/home/tester", ".config", "photosphere", "vault" });
    try std.testing.expectEqualStrings(expected, try vault_zig.plaintext_vault.DEFAULT_VAULT_DIR(allocator));
}

test "vault() exposes the PlaintextVault through the IVault interface" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var plaintext_vault = PlaintextVault.init(temp_dir.path);
    const vault = plaintext_vault.vault();

    try vault.set(allocator, io, .{ .name = "via-interface", .type = "plain", .value = "v" });
    const result = (try vault.get(allocator, io, "via-interface")).?;
    try std.testing.expectEqualStrings("v", result.value);
    const secrets = try vault.list(allocator, io);
    try std.testing.expectEqual(@as(usize, 1), secrets.len);
}

//
// Loads the secrets that generate.ts wrote to ts-vault/.
//
fn loadFixtureSecrets(allocator: std.mem.Allocator, io: std.Io) ![]ISecret {
    const json = try std.Io.Dir.cwd().readFileAlloc(io, "src/test/fixtures/secrets.json", allocator, .unlimited);
    return std.json.parseFromSliceLeaky([]ISecret, allocator, json, .{});
}

test "interop: Zig reads vault entries written by TypeScript" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const secrets = try loadFixtureSecrets(allocator, io);
    var vault = PlaintextVault.init("src/test/fixtures/ts-vault");

    for (secrets) |expected| {
        const result = (try vault.get(allocator, io, expected.name)).?;
        try std.testing.expectEqualStrings(expected.name, result.name);
        try std.testing.expectEqualStrings(expected.type, result.type);
        try std.testing.expectEqualStrings(expected.value, result.value);
    }

    const listed = try vault.list(allocator, io);
    try std.testing.expectEqual(secrets.len, listed.len);
    const sorted_expected = try allocator.dupe(ISecret, secrets);
    std.mem.sort(ISecret, sorted_expected, {}, secretNameLessThan);
    std.mem.sort(ISecret, listed, {}, secretNameLessThan);
    for (sorted_expected, listed) |expected, actual| {
        try std.testing.expectEqualStrings(expected.name, actual.name);
        try std.testing.expectEqualStrings(expected.value, actual.value);
    }
}

test "interop: Zig writes vault files byte-identical to TypeScript (so TypeScript reads them)" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const secrets = try loadFixtureSecrets(allocator, io);
    var temp_dir = try makeTempDir(allocator);
    defer temp_dir.remove();
    var vault = PlaintextVault.init(temp_dir.path);

    for (secrets) |secret| {
        try vault.set(allocator, io, secret);
    }

    var fixture_dir = try std.Io.Dir.cwd().openDir(io, "src/test/fixtures/ts-vault", .{ .iterate = true });
    defer fixture_dir.close(io);
    var iterator = fixture_dir.iterate();
    var file_count: usize = 0;
    while (try iterator.next(io)) |entry| {
        const expected = try fixture_dir.readFileAlloc(io, entry.name, allocator, .unlimited);
        const actual = try temp_dir.tmp_dir.dir.readFileAlloc(io, entry.name, allocator, .unlimited);
        try std.testing.expectEqualStrings(expected, actual);
        file_count += 1;
    }
    try std.testing.expectEqual(secrets.len, file_count);
}
